numpy/distutils/ccompiler.py

```diff
@@ -191,7 +230,8 @@ def CCompiler_compile(self, sources, output_dir=None, macros=None,
 
     def single_compile(args):
         obj, (src, ext) = args
-        self._compile(obj, src, ext, cc_args, extra_postargs, pp_opts)
+        if _needs_build(obj):
+            self._compile(obj, src, ext, cc_args, extra_postargs, pp_opts)
```

Comment:
hm I though I tested scipy with this, apparently I forgot to run the tests

checking the arguments is simple enough, would be nicer to add the extension name to the dependency filename instead but it is probably clunky to get that from the compiler object.